### PR TITLE
Add tagset module for encoding/decoding the tagset format

### PR DESCRIPTION
**Pull request by brettlangdon**
## Commit Message
{{title}}

**Purpose:**

The tagset format is used to encode arbitrary key/value pairs into a single string. This encoded value can then be passed along in HTTP headers/context propagation. This will initially be used to share tracer internal state between services.

**Format:**

```ebpf
tagset = tag, { ",", tag };
tag = ( identifier - space ), "=", identifier;
identifier = { ? ASCII 32-126 ? - equal or comma };\
equal or comma = "=" | ",";
space = " ";
```

**Examples:**

```python
>>> decode_tagset_string("key=value")
{"key": "value"}
>>> decode_tagset_string("a=1,b=2,c=3")
{"a": "1", "b": "2", "c": "3"}

>>> encode_tagset_values({"key": "value"})
"key=value"
>>> encode_tagset_values({"a": "1", "b": "2", "c": "3"})
"a=1,b=2,c=3"
```

**Performance:**

Cython vs Pure-Python implementations.

| Test | Cython | Python | Diff |
| -- | -- | -- | -- |
| Very large header | 33.7 us | 318 us | 89% faster|
| Large header | 18.2 us | 156 us | 88% faster |
| Medium header | 8.72 us | 80.5 us | 89% faster |
| **Small header** | 2.82 us | 31.8 us | 91% faster |
| **Empty header** | 59.2 ns | 128 ns | 53% faster |

### CODE REVIEW

**Review by Kyle-Verhoog — COMMENTED** [review]
I'd like to see some motivation for writing this in Cython vs pure python. I'd like to avoid Cython as much as possible for maintainability reasons. Since we're really just dealing with Python objects here I'm skeptical that the wins are going to be _that_ big (just saving the byte code evaluation) but I could be wrong

**Comment by brettlangdon** [comment]
Oh sorry, yeah I should have included performance numbers 🤦🏻 (tl;dr; it is 4x faster just having this in Cython, having cdefs for the local variables, etc etc)

**Comment by Kyle-Verhoog** [comment]
Do we have the magnitudes as well? I'd be willing to take 4x difference if it's 0.1us vs 0.4us when creating a span is ~20us as a for instance.

**Comment by brettlangdon** [comment]
The most common results will be small and empty header. Medium header is likely in the near future, large and very large are unlikely, but let's not say they would never happen.

This would be the impact on decoding a header value when extracting context propagation. This would be the impact extracting for every request. (Have not yet measured injecting/encoding performance).

| Test | Cython | Python | Diff |
| -- | -- | -- | -- |
| Very large header | 33.7 us | 318 us | 89% faster|
| Large header | 18.2 us | 156 us | 88% faster |
| Medium header | 8.72 us | 80.5 us | 89% faster |
| **Small header** | 2.82 us | 31.8 us | 91% faster |
| **Empty header** | 59.2 ns | 128 ns | 53% faster |

<details>
<summary>Perf test code</summary>

```python
import random
import string
import sys

import pyperf

from ddtrace.internal._tagset import TagsetDecodeError
from ddtrace.internal._tagset import decode_tagset_string


def is_equal(c):
    # '='
    return c == 61


def is_comma(c):
    # ','
    return c == 44


def is_valid_key_char(c):
    # string.printable - " ,="
    # 32 = " "
    # 44 = ",""
    # 61 = "="
    return (33 <= c <= 43) or (45 <= c <= 60) or (62 <= c <= 126)


# Same as is_valid_key_char except spaces are allowed
def is_valid_value_char(c):
    # string.printable - ",="
    # 44 = ",""
    # 61 = "="
    return c == 32 or is_valid_key_char(c)


def python_decode_tagset_string(tagset):
    # type: (str) -> Dict[str, str]
    res = {}
    cur_val = ""
    cur_key = ""
    is_parsing_key = True

    # No tagset provided, short circuit the response
    if not tagset:
        return res

    # DEV: Parse in a single pass of `tagset`
    #      `is_parsing_key` is used to know if we are on
    #      right or left side of an `=`
    for c in tagset:
        o = ord(c)
        if is_comma(o):
            # Allow leading/trailing commas without key/value pairs
            if is_parsing_key and not cur_key:
                continue

        if is_parsing_key:
            if is_equal(o):
                if not cur_key:
                    raise TagsetDecodeError("Empty keys are not allowed: {!r}".format(tagset))
                is_parsing_key = 0
                continue

            if not is_valid_key_char(o):
                raise TagsetDecodeError("Unexpected {!r} character for key {}: {!r}".format(c, cur_key, tagset))
            cur_key += c
        else:
            if is_comma(o):
                cur_val = cur_val.strip(" ")
                if not cur_val:
                    raise TagsetDecodeError("Empty values are not allowed: {!r}".format(tagset))
                res[cur_key] = cur_val
                cur_key, cur_val, is_parsing_key = "", "", 1
                continue

            if not is_valid_value_char(o):
                raise TagsetDecodeError(
                    "Unexpected character {!r} for value {}={}: {!r}".format(c, cur_key, cur_val, tagset)
                )
            cur_val += c

    # Reached EOF, do we have a key/value pair we need to save?
    if cur_key:
        cur_val = cur_val.strip(" ")
        if cur_val:
            res[cur_key] = cur_val
        else:
            raise TagsetDecodeError("Expected value for key {!r} instead got EOF: {!r}".format(cur_key, tagset))

    return res


def rands(size=6, chars=string.ascii_lowercase + string.digits):
    return "".join(random.choice(chars) for _ in range(size))


def gen_header(entries):
    header = []
    for _ in range(entries):
        header.append("{}={}".format(rands(size=24), rands(size=32)))

    return ",".join(header)


def add_args(args, cmd):
    cmd.rigorous = True


runner = pyperf.Runner(warmups=10, add_cmdline_args=add_args)
for (prefix, func) in [("cython", decode_tagset_string), ("python", python_decode_tagset_string)]:
    runner.bench_func("{}_very_large_header".format(prefix), func, gen_header(12))
    runner.bench_func("{}_large_header".format(prefix), func, gen_header(6))
    runner.bench_func("{}_medium_header".format(prefix), func, gen_header(3))
    runner.bench_func("{}_small_header".format(prefix), func, gen_header(1))
    runner.bench_func("{}_empty_header".format(prefix), func, "")
```
</details>

<details>
<summary>Raw results</summary>

```
cython_very_large_header: Mean +- std dev: 33.7 us +- 1.5 us
cython_large_header: Mean +- std dev: 18.2 us +- 1.5 us
cython_medium_header: Mean +- std dev: 8.72 us +- 0.84 us
cython_small_header: Mean +- std dev: 2.82 us +- 0.20 us
cython_empty_header: Mean +- std dev: 59.2 ns +- 6.8 ns
python_very_large_header: Mean +- std dev: 318 us +- 12 us
python_large_header: Mean +- std dev: 156 us +- 5 us
python_medium_header: Mean +- std dev: 80.5 us +- 4.5 us
python_small_header: Mean +- std dev: 31.8 us +- 7.2 us
python_empty_header: Mean +- std dev: 128 ns +- 3 ns
```

</details>

**Comment by brettlangdon** [comment]
There are failing tests because of iterating `dict.items()` which does not have guaranteed order. Will investigate/consider using an `OrderedDict` just for tests.

**Review by P403n1x87 — APPROVED** [review]

File: ddtrace/internal/_tagset.pyx [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        "a=1,b=2,c=3"
+
+    :param dict values: Dict[str,str] of key/value pairs to encode
+    :param int max_size: The max size to allow the resulting string to be (default: 512)
+    :rtype: str
+    :returns: tagset encoded key/value pairs
+    :raises TagsetMaxSizeError: Raised when we will exceed the provided max size
+    :raises TagsetEncodeError: Raised when we encounter an exception character in a key or value
+    """
+    # Allow subclasses of `dict` like `collections.OrderedDict`
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/internal/_tagset.pyx:162]: I think you are right that we can remove.

For context: before we had `dict values` as the parameter and now we have `object values`, so adding this runtime type check was mostly a way to persist the type checking that Cython was doing previously with `dict values`.

File: ddtrace/internal/_tagset.pyx [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    if not tagset:
+        return res
+
+    # DEV: Parse in a single pass of `tagset`
+    #      `is_parsing_key` is used to know if we are on
+    #      right or left side of an `=`
+    for c in tagset:
+        o = ord(c)
+        if is_comma(o):
+            # Allow leading/trailing commas without key/value pairs
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/internal/_tagset.pyx:106]: we can remove to reduce complexity

**Review by mabdinur — APPROVED** [review]

**Pull request merged**